src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_fc_reshape_to_weights.cpp

```diff
@@ -0,0 +1,102 @@
+    };
+    auto reshape_const_m = wrap_type<ov::opset1::Constant>(consumers_count(1));
+    auto reshape_m = wrap_type<ov::opset1::Reshape>({mul_m, reshape_const_m}, one_consumer_rank_2);
+
+    auto transpose_const_m = wrap_type<ov::opset1::Constant>();
+    auto transpose_m = wrap_type<ov::opset1::Transpose>({reshape_m, transpose_const_m});
+    auto weights_input_m = std::make_shared<ov::pass::pattern::op::Or>(ov::OutputVector{reshape_m, transpose_m});
+
+    auto data_m = any_input();
+    auto fully_connected_m = wrap_type<ov::intel_cpu::FullyConnectedNode>({data_m, weights_input_m});
+
+    ov::matcher_pass_callback callback = [&](ov::pass::pattern::Matcher& m) {
+        const auto fully_connected = m.get_match_root();
+        const auto weights_path = fully_connected->get_input_node_shared_ptr(1);
+        const bool with_transpose = ov::is_type<ov::opset1::Transpose>(weights_path);
+        if (with_transpose) {
+            const auto transpose_const = ov::as_type_ptr<ov::opset1::Constant>(weights_path->get_input_node_shared_ptr(1));
+            if (transpose_const->cast_vector<int>() != std::vector<int>{1, 0}) {
+                return false;
+            }
+        }
+
+        const auto& fc_input_shape = fully_connected->get_input_shape(1);
+        const auto reshape = with_transpose ? weights_path->get_input_node_shared_ptr(0) : weights_path;
+
+        auto check_decompression_const = [&](const std::shared_ptr<ov::Node>& node) {
+            if (!ov::is_type<ov::opset1::Constant>(node))
+                return false;
+            ov::Shape expected_shape(3, 1);
+            const size_t out_channels_idx = with_transpose ? 2 : 1;
+            expected_shape[out_channels_idx] = fc_input_shape[0];
+            return node->get_output_shape(0) == expected_shape;
+        };
+
+        const auto mul = reshape->get_input_node_shared_ptr(0);
+        if (!check_decompression_const(mul->get_input_node_shared_ptr(1)))
+            return false;
+        const auto mul_parent = mul->get_input_node_shared_ptr(0);
+        const bool with_subtract = ov::is_type<ov::opset1::Subtract>(mul_parent);
+        if (with_subtract && !check_decompression_const(mul_parent->get_input_node_shared_ptr(1)))
+            return false;
+
+        const auto convert = with_subtract ? mul_parent->get_input_node_shared_ptr(0) : mul_parent;
+        const auto weights = convert->get_input_node_shared_ptr(0);
+        ov::Shape expected_weights_shape(3, 1);
+        expected_weights_shape[1] = fc_input_shape[with_transpose ? 1 : 0];
+        expected_weights_shape[2] = fc_input_shape[with_transpose ? 0 : 1];
+        if (weights->get_output_shape(0) != expected_weights_shape)
+            return false;
+
+        auto squeeze_constant = [](const std::shared_ptr<ov::Node>& node) {
+            const auto constant = ov::as_type_ptr<ov::opset1::Constant>(node);
+            auto shape = constant->get_shape();
+            OPENVINO_ASSERT(shape[0] == 1);
```

Comment:
usually we try to avoid using exceptions/asserts in the transformations code.
can we check the shapes first and if they contain shape[0] == 1, then do not apply the transformation at all and return false immediately?

Reply:
Actually we already check it upper so I just removed this useless assert